### PR TITLE
feat: run extra query on QueryObject and add compare operator for post_processing

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

This PR is part of the `advanced analytics`. in this PR introduces:
- `run extra query` on the v1 data API
- add `compare` operator for time compare
- refactoring `get_df_payload`
- persistent extra query results(per time offset)
- unit test for these new functions
- minor improvements in `Makefile`, `diff operator` and `pylint`

client-side codes at: https://github.com/apache-superset/superset-ui/pull/1170

### Preparing test data
please use `random_time_series` dataset for this test.

### Calculation changes (there are 3 changes)
Now the time offset calculation has some changes from before. 

1. Extra query `where clause` offset calculation, use **time units(x years/x months/x weeks)** as calculation offset instead of **day timedelta** to calculate offsets. This will avoid leap year with incorrect calculation

Before **Main** query(line chart) 
```
SELECT DATE_TRUNC('month', ds) AS __timestamp,
       COUNT(*) AS count
FROM random_time_series
WHERE ds >= TO_TIMESTAMP('2016-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
  AND ds < TO_TIMESTAMP('2017-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
GROUP BY DATE_TRUNC('month', ds)
ORDER BY count DESC
LIMIT 50000
```

Befor **1 year ago offset** query(line chart) 
```
SELECT DATE_TRUNC('month', ds) AS __timestamp,
       COUNT(*) AS count
FROM random_time_series
WHERE ds >= TO_TIMESTAMP('2015-03-02 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
  AND ds < TO_TIMESTAMP('2016-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
GROUP BY DATE_TRUNC('month', ds)
ORDER BY count DESC
LIMIT 50000
```

After **Main** query (time-series chart)
```
SELECT DATE_TRUNC('month', ds) AS __timestamp,
       COUNT(*) AS count
FROM random_time_series
WHERE ds >= TO_TIMESTAMP('2016-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
  AND ds < TO_TIMESTAMP('2017-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
GROUP BY DATE_TRUNC('month', ds)
LIMIT 50000
```

After **1 year ago offset** query
```
SELECT DATE_TRUNC('month', ds) AS __timestamp,
       COUNT(*) AS count
FROM random_time_series
WHERE ds >= TO_TIMESTAMP('2015-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
  AND ds < TO_TIMESTAMP('2016-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
GROUP BY DATE_TRUNC('month', ds)
LIMIT 50000
```

2. Use `join on __timestamp` instead of `concat` merge main dataframe and extra dataframe. This ensures that the main query and the extension query have the same time granularity calculation on metrics.

Consider this scenario(time-series chart):
Screen Shot 2021-07-01 at 1 10 50 PM

The main query is:
```
SELECT DATE_TRUNC('month', ds) AS __timestamp,
       COUNT(*) AS count
FROM random_time_series
WHERE ds >= TO_TIMESTAMP('2016-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
  AND ds < TO_TIMESTAMP('2017-03-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
GROUP BY DATE_TRUNC('month', ds)
LIMIT 50000
```

with the time shift `28 days ago` query is:
```
SELECT DATE_TRUNC('month', ds) AS __timestamp,
       COUNT(*) AS count
FROM random_time_series
WHERE ds >= TO_TIMESTAMP('2016-02-02 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
  AND ds < TO_TIMESTAMP('2017-02-01 00:00:00.000000', 'YYYY-MM-DD HH24:MI:SS.US')
GROUP BY DATE_TRUNC('month', ds)
LIMIT 50000
```

It is obvious that the extra query is missing the data of **2016-02-01**(the main query grain is month, but the extra query is missing a day, the metric is incorrect), when **main dataframe left join extra dataframe on __timestamp**, the extra data that is not "align" on the main dataframe will be removed automatically.

extra query dataframe
image

merged dataframe
image

3. Not allowed `x time unit` format for timeoffset. Must be specified `x timeunit ago` or `x timeunit later`

#### A. prepare code
1. pull associated PR in `superset-ui`: https://github.com/apache-superset/superset-ui/pull/1170 
2. pull current pr in `superset`
3. the directory like this:
```bash
$ tree -L 1
.
├── superset-ui
└── superset
```

#### B. Build superset-ui
run after commands in terminal
```bash
$ cd superset-ui
superset-ui$ yarn clean
superset-ui$ rm -rf ./{packages,plugins}/*/node_modules ./node_modules
superset-ui$ yarn clean-npm-lock

superset-ui$ yarn
superset-ui$ yarn build
```

#### C. Build superset-frontend
```
$ cd superset/superset-frontend
superset-frontend$ npm ci
superset-frontend$ npm link --legacy-peer-deps ../../superset-ui/plugins/plugin-chart-echarts/ ../../superset-ui/packages/superset-ui-chart-controls/ ../../superset-ui/packages/superset-ui-core/
```

#### D. Hack core module
1. change `package.json` in superset-ui-core
```
diff --git a/packages/superset-ui-core/package.json b/packages/superset-ui-core/package.json
index 30a78f1e..a67dfd40 100644
--- a/packages/superset-ui-core/package.json
+++ b/packages/superset-ui-core/package.json
@@ -32,8 +32,6 @@
   },
   "dependencies": {
     "@babel/runtime": "^7.1.2",
-    "@emotion/cache": "^11.1.3",
-    "@emotion/react": "^11.1.5",
     "@emotion/styled": "^11.3.0",
     "@types/d3-format": "^1.3.0",
     "@types/d3-interpolate": "^1.3.1",
@@ -65,6 +63,8 @@
     "@types/react": "*",
     "@types/react-loadable": "*",
     "react": "^16.13.1",
-    "react-loadable": "^5.5.0"
+    "react-loadable": "^5.5.0",
+    "@emotion/cache": "^11.1.3",
+    "@emotion/react": "^11.1.5"
   }
 }
```
2. build each package of superset-ui
```
$ cd superset-ui/packages/superset-ui-core/
superset-ui-core$ yarn

$ cd superset-ui/packages/superset-ui-chart-controls/
superset-ui-chart-controls$ yarn

$ cd superset-ui/plugins/plugin-chart-echarts/
plugin-chart-echarts$ yarn
```

#### E. run dev server in Superset
```
$ cd superset/superset-frontend
superset-frontend$ npm run dev-server
```

### TESTING INSTRUCTIONS

Added UT in python codebase

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache-superset/superset-ui/pull/1170
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/common/query_object.py [hunk 1]
```diff
@@ -94,6 +96,7 @@ class QueryObject:
     datasource: Optional[BaseDatasource]
     result_type: Optional[ChartDataResultType]
     is_rowcount: bool
+    time_offset: List[str]
```
Comment (villebro) [inline superset/common/query_object.py:99]: I think it would be great to deprecate the current string-based offset, and use a periods-based approach, much like in rolling window, using the time grain as the time offset unit.

File: superset/common/query_context.py [hunk 2]
```diff
@@ -97,6 +101,62 @@ def __init__(  # pylint: disable=too-many-arguments
...
+                query_object_clone.from_dttm = get_past_or_future(
+                    offset, outer_from_dttm,
+                )
+                query_object_clone.to_dttm = get_past_or_future(offset, outer_to_dttm,)
+            except ValueError as ex:
+                raise QueryObjectValidationError(str(ex))
+            # make sure subquery use main query where clause
+            query_object_clone.inner_from_dttm = outer_from_dttm
+            query_object_clone.inner_to_dttm = outer_to_dttm
+            query_object_clone.time_offset = []
```
Comment (villebro) [inline superset/common/query_context.py:125]: I wonder if we should add `time_offset` to the `QueryObject` schema and rename the current one to `time_offsets`, adding both to the cache key. Example:

We want to make a query with two offsets: one year ago and two years ago.

The "actual" main query that gets executed and cached (no additional columns added yet):
```python
time_offsets: None
time_offset: None
```

First extra query (gets concatenated to the previous dataframe):
```python
time_offsets: None
time_offset: 1
```

Second extra query (also concatenated to the main dataframe):
```python
time_offsets: None
time_offset: 2
```


Main `QueryObject`:
```python
time_offsets: [1, 2]
time_offset: None
```

Finally, when the full query object is constructed, the following result would be cached with the following keys:

This way the main query result would be persisted along with the results of the extra query results without the need to rebuild the full dataframe on each request, and the extra queries could then also be cached individually.

**Review by villebro — COMMENTED** [review]
Some comments. Also, I'm unable to get the offsets to work. When I try to do a 1 year offset, I get the following results (notice that the offset label is correct, but the data is the same as the original query):
image

The same parameters on the NVD3 Line chart:
image

**Comment by zhaoyongjie** [comment]
> Some comments. Also, I'm unable to get the offsets to work. When I try to do a 1 year offset, I get the following results (notice that the offset label is correct, but the data is the same as the original query):
> image
> 
> The same parameters on the NVD3 Line chart:
> image


This is because the original offset is always negative. (offset backward, if the current is 2016-2017, `the 1 year offset ago`, is 2017-2018)

For the new design, I would like to put `ago/later` into the time interval. This allows the user to know which direction to do offset.

for this question, the dataset may be have not before `1 year ago` data in the new time-series viz.

File: superset/utils/pandas_postprocessing.py [hunk 3]
```diff
@@ -425,9 +429,11 @@ def select(
 
 
 @validate_column_args("columns")
-def diff(df: DataFrame, columns: Dict[str, str], periods: int = 1,) -> DataFrame:
+def diff(
+    df: DataFrame, columns: Dict[str, str], periods: int = 1, axis: int = 0,
```
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset/utils/pandas_postprocessing.py:433]: I also often have trouble remembering. Then I used 1 for "vertical" to remember that 1 for "columns".

File: superset/utils/pandas_postprocessing.py [hunk 4]
```diff
@@ -436,14 +442,68 @@ def diff(df: DataFrame, columns: Dict[str, str], periods: int = 1,) -> DataFrame
...
+    df_diff = df_diff.diff(periods=periods, axis=axis)
     return _append_columns(df, df_diff, columns)
 
 
+@validate_column_args("source_columns", "compare_columns")
+def compare(
+    df: DataFrame,
+    source_columns: List[str],
+    compare_columns: List[str],
+    compare_type: Optional[str],
```
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset/utils/pandas_postprocessing.py:459]: nice catch!

File: superset/common/query_context.py [hunk 5]
```diff
@@ -130,11 +191,16 @@ def get_query_result(self, query_object: QueryObject) -> Dict[str, Any]:
             if self.enforce_numerical_metrics:
                 self.df_metrics_to_num(df, query_object)
 
+            if query_object.time_offset:
+                df, offset_sql = self.processing_time_offset(df, query_object)
+                query += ";\n\n".join(offset_sql)
+                query += ";\n\n"
```
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset/common/query_context.py:197]: The purpose of this is to ensure that all SQLs are followed by a semicolon.
```
main query;

offset1 query;

offset2 query;

offset3 query;
             ^ append semicolon at end

```

File: superset/common/query_context.py (hunk 2 above)
Comment (zhaoyongjie) [reply superset/common/query_context.py]: I do some refactoring for the function of `get_df_payload`. Now the main query and the extra query should be cached.

**Comment by zhaoyongjie** [comment]
> Some comments. Also, I'm unable to get the offsets to work. When I try to do a 1 year offset, I get the following results (notice that the offset label is correct, but the data is the same as the original query):
> image
> 
> The same parameters on the NVD3 Line chart:
> image

This bug has been fixed.

image

File: superset/charts/commands/exceptions.py [hunk 6]
```diff
@@ -56,6 +56,23 @@ def __init__(self, human_readable: str) -> None:
         )
 
 
+class TimeDeltaUnclearError(ValidationError):
+    """
+    Time delta is in valid error.
```
Comment (mistercrunch) [inline superset/charts/commands/exceptions.py:61]: typo "invalid", better use "unclear" for consistency

File: superset/common/query_context.py [hunk 7]
```diff
@@ -101,21 +104,143 @@ def __init__(  # pylint: disable=too-many-arguments
...
-        """Returns a pandas dataframe based on the query object"""
+    @staticmethod
+    def left_join_on_dttm(
+        left_df: pd.DataFrame, right_df: pd.DataFrame
+    ) -> pd.DataFrame:
+        df = left_df.set_index(DTTM_ALIAS).join(right_df.set_index(DTTM_ALIAS))
+        df.reset_index(level=0, inplace=True)
+        return df
+
+    def processing_time_offsets(
```
Comment (mistercrunch) [inline superset/common/query_context.py:115]: NIT: looking at the function signature, personally I'm not big on methods that return a tuple of mixed objects, it feels like maybe we should be mutating the object this method is part of, or returning a more complex object with a proper name "CachedDataFrame" or something (unclear to me at this point what the best name would be). I namedtuple could be a good compromise.

I'm not sure what it implies as I'm just starting to review this code

File: superset/utils/date_parser.py [hunk 8]
```diff
@@ -106,6 +119,16 @@ def dttm_from_timetuple(date_: struct_time) -> datetime:
     )
 
 
+def get_past_or_future(
+    human_readable: Optional[str], source_time: Optional[datetime] = None,
+) -> datetime:
+    cal = parsedatetime.Calendar()
```
Comment (mistercrunch) [inline superset/utils/date_parser.py:125]: The next few lines have a lot in common with `parse_human_timedelta`, should we refactor them both use a common simple method?

File: superset/common/utils.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+stats_logger: BaseStatsLogger = config["STATS_LOGGER"]
+logger = logging.getLogger(__name__)
+
+_cache: Dict[CacheRegion, Cache] = {
+    CacheRegion.DEFAULT: cache_manager.cache,
+    CacheRegion.DATA: cache_manager.data_cache,
+}
+
+
+class QueryCacheManager:
```
Comment (mistercrunch) [inline superset/common/utils.py:42]: docstring pls, I'm not sure what exactly this is for

**Review by mistercrunch — COMMENTED** [review]
Posting a partial first pass as I have limited time to do a more thorough review. I asked @betodealmeida to help review this too.

File: superset/charts/commands/exceptions.py (hunk 6 above)
Comment (zhaoyongjie) [reply superset/charts/commands/exceptions.py:61]: Nice catch, thank you.

File: superset/utils/date_parser.py (hunk 8 above)
Comment (zhaoyongjie) [reply superset/utils/date_parser.py:125]: This function returns `datetime`, while the following ones return `timedelta`.

File: superset/common/query_context.py (hunk 7 above)
Comment (zhaoyongjie) [reply superset/common/query_context.py:115]: I'm going to do a little refactoring for the return value(Returns an object for this function). Thanks for the heads up.

File: superset/common/query_context.py [hunk 10]
```diff
@@ -101,21 +104,143 @@ def __init__(  # pylint: disable=too-many-arguments
...
+    ) -> Tuple[pd.DataFrame, List[str], List[Optional[str]]]:
+        # ensure query_object is immutable
+        query_object_clone = copy.copy(query_object)
+        rv_sql = []
+        cache_keys = []
+
+        time_offsets = query_object.time_offsets
+        outer_from_dttm = query_object.from_dttm
+        outer_to_dttm = query_object.to_dttm
+        for offset in time_offsets:
```
Comment (ktmud) [inline superset/common/query_context.py:126]: I'm not sure you need to run and cache a completely new query for each offset.

Can we somehow compute the final time periods and generate proper `WHERE` conditions with `or` filters instead?

```python
def get_time_periods_for_offsets(time_range, offsets):
    [start, end] = time_range
    periods = [time_range]
    for offset in periods:
        periods.append([start += offset, end += offset])
    return periods
```

Then change https://github.com/apache/superset/blob/bee386e643a202f182a5cffd03c460fc27efe959/superset/connectors/sqla/models.py#L1370-L1375

to something like

```python
inner_time_filter = or_([dttm_col.between(start, end) for start, end in periods])
subq = subq.where(and_(*(where_clause_and + [inner_time_filter]))
```

**Review by ktmud — COMMENTED** [review]
I have some reservation on running multiple queries for each period after offset----it basically means an additional 1x query time for each new offset---plus the overheads of transferring potentially duplicate rows with overlapping periods.

File: superset/common/utils.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        datasource_uid: Optional[str] = None,
+        region: CacheRegion = CacheRegion.DEFAULT,
+    ) -> None:
+        value = {
+            "df": self.df,
+            "query": self.query,
+            "annotation_data": self.annotation_data,
+        }
+        if self.is_loaded and key and self.status != QueryStatus.FAILED:
+            self.set(key, value, timeout, datasource_uid, region)
```
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset/common/utils.py:107]: nice catch! thanks

File: superset/utils/core.py [hunk 12]
```diff
@@ -115,6 +115,8 @@
 
 DTTM_ALIAS = "__timestamp"
 
+TIME_COMPARISION = "__"
```
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset/utils/core.py:118]: This separator is reserved for now because it has to match the frontend.

File: superset/utils/date_parser.py (hunk 8 above)
Comment (zhaoyongjie) [reply superset/utils/date_parser.py:125]: I will open a separate PR to fix these problems.

File: superset/common/query_context.py (hunk 10 above)
Comment (zhaoyongjie) [reply superset/common/query_context.py:133]: For where clause combined by `or` operator, I estimate that the system consumption is approximately equal to multiple queries. This is because the `or` operator does not reduce rows scan for the database engine. And we don't have the opportunity to cache time offset. Let me explain.

### Use `or` operator in the where clause
- unable to cache every timeoffset slice
- unable to easy generate finial dataframe

image



### Use extra query
image

Comment (ktmud) [reply superset/common/query_context.py:133]: ＞ This is because the or operator does not reduce rows scan for the database engine. 

It would reduce rows scanned if there are significant overlaps among the offset time periods. E.g. you query for two years of data and offset by 1 year.

> unable to easily generate the final dataframe

Isn't each sub-dataframe a between(start_time, end_time) filter on the query result dataframe? We should probably use pandas to handle the time periods and join by datetime index anyway, if we are not already doing that, so the split & join by time process shouldn't be that hard, either.

**Comment by suddjian** [comment]
Tested the new changes and it looks like it's all working well! Gonna spend some time reviewing the code next.

File: .pylintrc [hunk 13]
```diff
@@ -81,7 +81,7 @@ confidence=
...
+disable=long-builtin,dict-view-method,intern-builtin,suppressed-message,no-absolute-import,unpacking-in-except,apply-builtin,delslice-method,indexing-exception,old-raise-syntax,print-statement,cmp-builtin,reduce-builtin,useless-suppression,coerce-method,input-builtin,cmp-method,raw_input-builtin,nonzero-method,backtick,basestring-builtin,setslice-method,reload-builtin,oct-method,map-builtin-not-iterating,execfile-builtin,old-octal-literal,zip-builtin-not-iterating,buffer-builtin,getslice-method,metaclass-assignment,xrange-builtin,long-suffix,round-builtin,range-builtin-not-iterating,next-method-called,parameter-unpacking,unicode-builtin,unichr-builtin,import-star-module-level,raising-string,filter-builtin-not-iterating,using-cmp-argument,coerce-builtin,file-builtin,old-division,hex-method,missing-docstring,too-many-lines,ungrouped-imports,import-outside-toplevel,raise-missing-from,super-with-arguments,bad-option-value,too-few-public-methods,too-many-locals,too-many-arguments,too-many-statements
```
Comment (suddjian) [inline .pylintrc:84]: nice

File: superset/utils/pandas_postprocessing.py [hunk 14]
```diff
@@ -357,8 +359,10 @@ def rolling(  # pylint: disable=too-many-arguments
     rolling_type_options = rolling_type_options or {}
     df_rolling = df[columns.keys()]
     kwargs: Dict[str, Union[str, int]] = {}
-    if not window:
+    if not window and window != 0:
```
Comment (suddjian) [inline superset/utils/pandas_postprocessing.py:362]: I believe this is equivalent? Unless there is some python silliness I'm not familiar with.

```suggestion
    if window is None:
```

**Comment by suddjian** [comment]
Code changes LGTM, but I'm not as confident with this code as I'd like to be. Curious how @villebro and @ktmud feel about the changes to the query behavior.

File: superset/utils/pandas_postprocessing.py (hunk 14 above)
Comment (zhaoyongjie) [reply superset/utils/pandas_postprocessing.py:362]: nice catch! I will change type of `window`

File: .pylintrc (hunk 13 above)
Comment (ktmud) [reply .pylintrc]: I wouldn't recommend making this change. We already have too many gigantic files/functions, even though sometimes not very actionable, having the disabling rule as inline comments serves as a reminder that certain piece of code needs refactoring.

File: superset/utils/core.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (villebro) [reply superset/utils/core.py:118]: > I have some reservation on running multiple queries for each period after offset----it basically means an additional 1x query time for each new offset---plus the overheads of transferring potentially duplicate rows with overlapping periods.

TL;DR: if we want to maintain feature parity with the current time offset functionality in `viz.py` (which issues multiple queries), we need to do that here, too.

@ktmud This was my initial reaction, too, and I assumed we could just add the offsets by based on the initial dataframe. However, after studying how this currently works, I noticed that we need to issue separate queries for each offset, as they will retrieve data for different time ranges. Take this screenshot from @zhaoyongjie 's comment above:
image
Here you can see that the "1 year ago" offset in year 1980 is in fact the data for 1979, which isn't visible in the original series. If we wanted to support arbitrary offsets based on just one query response, we would need to know the maximum offset, query based on that, and then truncate the original series to it's original time ranges etc.

File: superset/common/query_context.py (hunk 10 above)
Comment (villebro) [reply superset/common/query_context.py:133]: > ＞ This is because the or operator does not reduce rows scan for the database engine.
> 
> It would reduce rows scanned if there are significant overlaps among the offset time periods. E.g. you query for two years of data and offset by 1 year.
> 
> > unable to easily generate the final dataframe
> 
> Isn't each sub-dataframe a between(start_time, end_time) filter on the query result dataframe? We should probably use pandas to handle the time periods and join by datetime index anyway, if we are not already doing that, so the split & join by time process shouldn't be that hard, either.

I believe moving this type of logic into Superset would be a slippery slope to introducing logic that's usually best handled by the analytical database and cause major maintainability overhead. I'm open to considering this down the road, but it would require some more discussion to ensure we don't end up building a pseudo-database engine inside Superset. Maybe we can revisit this if/when we start working on adding the semantic layer for table joins?

File: .pylintrc (hunk 13 above)
Comment (villebro) [reply .pylintrc]: I agree; half the time I run into this it's not actionable, but the other half are highly actionable and are triggered on functions that have experienced multiple regressions over the years due to being overly complex.

File: .pylintrc [hunk 15]
```diff
@@ -90,6 +90,8 @@ disable=
     super-with-arguments,
     too-few-public-methods,
     too-many-locals,
+    too-many-arguments,
+    too-many-statements
```
Comment (villebro) [inline .pylintrc:94]: As per @ktmud 's recommendation, I'd also recommend keeping these intact.

File: superset/charts/commands/exceptions.py [hunk 16]
```diff
@@ -30,7 +30,7 @@
 
 class TimeRangeUnclearError(ValidationError):
     """
-    Time range is in valid error.
+    Time range is unclear error.
```
Comment (villebro) [inline superset/charts/commands/exceptions.py:33]: Not caused by this PR, but I think "Unclear" is slightly ambiguous. I think it could be more clear to call it `TimeRangeUnparseableError` for time ranges that are totally incorrect/unparseable, and have a separate `TimeRangeAmbiguousError` for time ranges that are ambiguous, e.g. '1 year' as this exception seems to be used now.

Btw, I noticed that there are unit tests for `parse_human_datetime` in the `test_parse_human_datetime` integration test, but the function already has `doctests` attached to it. If it's not a big deal, could we move all these tests into one place, either into the doctests or into the existing test function in `test_parse_human_datetime`?

File: superset/charts/commands/exceptions.py [hunk 17]
```diff
@@ -56,6 +56,23 @@ def __init__(self, human_readable: str) -> None:
...
 
+class TimeDeltaUnclearError(ValidationError):
+    """
+    Time delta is unclear error.
+    """
+
+    def __init__(self, human_readable: str) -> None:
+        super().__init__(
+            _(
+                "Time delta is unclear."
```
Comment (villebro) [inline superset/charts/commands/exceptions.py:67]:
```suggestion
                "Time delta is ambiguous."
```

File: superset/common/query_context.py [hunk 18]
```diff
@@ -101,21 +111,143 @@ def __init__(  # pylint: disable=too-many-arguments
...
+            if not query_object.from_dttm or not query_object.to_dttm:
+                raise QueryObjectValidationError(
+                    _(
+                        "An enclosed time range (both start and end) must be specified "
+                        "when using a Time Comparison."
+                    )
+                )
+            # `offset` is added to the hash function
+            cache_key = self.query_cache_key(query_object_clone, time_offset=offset)
+            _cache = QueryCacheManager.get(cache_key, CacheRegion.DATA, self.force)
```
Comment (villebro) [inline superset/common/query_context.py:156]: Is there a reason this is called `_cache` and not `cache` or `query_cache` or similar?

File: superset/common/query_context.py [hunk 19]
```diff
@@ -308,47 +436,16 @@ def get_annotation_data(self, query_obj: QueryObject) -> Dict[str, Any]:
...
-                logger.info("Serving from cache")
-
-        if force_cached and not is_loaded:
-            logger.warning(
-                "force_cached (QueryContext): value not found for key %s", cache_key
-            )
-            raise CacheLoadError("Error loading data from cache")
+        _cache = QueryCacheManager.get(
+            cache_key, CacheRegion.DATA, self.force, force_cached,
+        )
```
Comment (villebro) [inline superset/common/query_context.py:446]: Same question here about naming of `_cache`

File: superset/common/utils.py [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+_cache: Dict[CacheRegion, Cache] = {
+    CacheRegion.DEFAULT: cache_manager.cache,
+    CacheRegion.DATA: cache_manager.data_cache,
+}
+
+
+class QueryCacheManager:
+    """
+    Class for manage query-cache getting and setting
+    """
```
Comment (villebro) [inline superset/common/utils.py:45]: This is a great improvement to the readability of `QueryContext`!

**Review by villebro — COMMENTED** [review]
A few comments - but in general really great work and LGTM! Tested to work correctly in all use cases I could think of and ready to approve after you've had the opportunity to comment on the review comments.

File: .pylintrc (hunk 13 above)
Comment (zhaoyongjie) [reply .pylintrc]: removed these 2 rules.

File: .pylintrc (hunk 15 above)
Comment (zhaoyongjie) [reply .pylintrc:94]: removed these two rules.

File: superset/common/query_context.py (hunk 18 above)
Comment (zhaoyongjie) [reply superset/common/query_context.py:156]: Just for private variables. I will rename `_cache` to `cache`

File: superset/common/query_context.py (hunk 19 above)
Comment (zhaoyongjie) [reply superset/common/query_context.py:446]: same before

File: superset/charts/commands/exceptions.py (hunk 16 above)
Comment (zhaoyongjie) [reply superset/charts/commands/exceptions.py:33]: Nice! I have done these changes.

File: .pylintrc (hunk 13 above)
Comment (ktmud) [reply .pylintrc]: Note that it's possible to customize these rules: http://pylint.pycqa.org/en/latest/technical_reference/features.html?highlight=Max-args#design-checker-options

Instead of disabling them, maybe we can also update them to more sensible values that are the most suitable for Superset codebase.

**Review by villebro — APPROVED** [review]
LGTM! 🚀
